### PR TITLE
Netty: Call ClientTransport.Listener.transportReady() after receiving…

### LINKED ISSUE #829: Call ClientTransport.Listener.transportReady() in a more appropriate time.
Currently we call it after the TCP connection is connected, we should call it after receiving the settings frame, so that we know for sure that the server accepted the connection.

### PR DESCRIPTION
**Pull request by madongfly**
… SETTINGS frame.

Fixes #829.

Manually tested with Reconnect interop test:
Prior to this PR, the test can only pass when TLS is enabled, with this PR, the test passed for both TLS and non-TLS cases.

### CODE REVIEW

File: netty/src/main/java/io/grpc/netty/NettyClientTransport.java [hunk 1]
```diff
@@ -183,20 +184,6 @@ public void operationComplete(ChannelFuture future) throws Exception {
     }).channel();
     // Start the write queue as soon as the channel is constructed
     handler.startWriteQueue(channel);
-    // This write will have no effect, yet it will only complete once the negotiationHandler
-    // flushes any pending writes.
-    channel.write(NettyClientHandler.NOOP_MESSAGE).addListener(new ChannelFutureListener() {
-      @Override
-      public void operationComplete(ChannelFuture future) throws Exception {
-        if (future.isSuccess()) {
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyClientTransport.java]: I'd feel better if we left this for now and just did nothing if `future.isSuccess()` is true (a.k.a., just don't call `transportReady()`). I do think we _could_ remove it, but then the transport won't have a good status code of what went wrong (which doesn't matter _today_, but...)

File: netty/src/main/java/io/grpc/netty/NettyClientTransport.java [hunk 2]
```diff
@@ -255,7 +242,13 @@ private NettyClientHandler newHandler() {
     frameWriter = new Http2OutboundFrameLogger(frameWriter, frameLogger);
 
     BufferingHttp2ConnectionEncoder encoder = new BufferingHttp2ConnectionEncoder(
-            new DefaultHttp2ConnectionEncoder(connection, frameWriter));
+            new DefaultHttp2ConnectionEncoder(connection, frameWriter)) {
+      @Override
+      public ChannelFuture writeSettingsAck(ChannelHandlerContext ctx, ChannelPromise promise) {
+        listener.transportReady();
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyClientTransport.java]: We could receive many acks over the course of a connection. We should keep a boolean of whether or not we are already ready (so we don't call `transportReady()` multiple times).

Comment (madongfly) [inline netty/src/main/java/io/grpc/netty/NettyClientTransport.java]: I think this is called to write acks for a receiving SETTINGS frame, and a gRpc server only send one SETTINGS frame at the start of the connection, so we also only write one ack for it, right?

File: netty/src/main/java/io/grpc/netty/NettyClientTransport.java (hunk 1 above)
Comment (madongfly) [inline netty/src/main/java/io/grpc/netty/NettyClientTransport.java]: Done.

**Comment by madongfly** [comment]
@ejona86, PTAL, thanks!

File: netty/src/main/java/io/grpc/netty/NettyClientTransport.java (hunk 2 above)
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyClientTransport.java]: That assumes the remote implementation does not do something it is entitled to do, which is inappropriate. The server is fully entitled to send SETTINGS frames at any time.

At the moment [we do actually send an extra SETTINGS frame](https://github.com/grpc/grpc-java/blob/master/netty/src/main/java/io/grpc/netty/NettyServerHandler.java#L365) when connecting. That's because Netty sends one internally but didn't let us modify it before it was sent, so we had to send our own. That'll be resolved in the next Netty beta, though.

Comment (madongfly) [inline netty/src/main/java/io/grpc/netty/NettyClientTransport.java]: Done.

Will make another PR for OkHttp too.

**Comment by madongfly** [comment]
@ejona86 PTAL, thanks!

**Comment by ejona86** [comment]
@madongfly LGTM
